### PR TITLE
1/n [CI Build] - Refactor the build step to its own class and adding integration tests

### PR DESCRIPTION
**Pull request by pablo-code-org**
I have spent hours reading ci_builder, runner, and rake task code to not try to make it "simpler" for future engineers.

I want to refactor CI Builder into smaller class. 

The final goal is to have 4-5 classes that do the core elements for ci_build and add integration test for each.

#### Overview

This Pull Request introduces a series of enhancements to the Continuous Integration (CI) Builder module in our project. The CI Builder is a crucial part of our deployment pipeline, ensuring that every change committed to the codebase is built, tested, and verified before being deployed to staging or production environments. This PR aims to streamline the CI build process, improve its reliability, and make it easier to maintain.

#### Key Changes

1. **Integration of `CiBuildMetrics` Module:**  
   The `CiBuildMetrics` module has been incorporated to log build details to the metrics database. This module takes care of logging the build status (`:start`, `:success`, `:failed`) against the commit hash and environment. It currently operates only in the staging environment.

2. **Build Process Automation in `CiBuilder`:**
   The `CiBuilder` module has been updated to automate various steps in the build process. Key functionalities include:
   - **Build Initiation** 
   - **Repository Update:**
   - **Build History Logging:** 
   - **Dependency Installation:** 
   - **Honeybadger Notification:** 
   - **Execution of CI Tasks:** 

## Testing story

The current test suite covers essential components of the CI Builder, ensuring that critical functionalities like build initiation, updates, and task execution are validated. However, there are areas where the test coverage could be enhanced to ensure comprehensive validation.

#### Recommendations for Testing
1. **Error Handling:** More tests around error scenarios, ensuring the builder gracefully handles and logs unexpected exceptions.
2. **Dependency Installation:** Validate scenarios where dependencies fail to install, and how it impacts the build process.
3. **Environment-Specific Behaviors:** Tests to cover different behaviors in staging, production, and development environments.

This image is the output of the build process after running on an adhoc machine. 
- I verified the change was on the machine.
- I ran the build successfully. 

image

## PR Checklist:

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [x] New features are translatable or updates will not break translations
- [x] Relevant documentation has been added or updated
- [x] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [x] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: aws/ci_builder.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  def build_required?
+    RakeUtils.git_updates_available? || File.file?(BUILD_STARTED) || !CDO.daemon
+  end
+
+  def update_repository
+    RakeUtils.git_fetch
+    count = RakeUtils.git_update_count
+    RakeUtils.git_pull if count > 0
+    [1, count].max # I think this is not needed. unless count can be a real number between 0 and 1.
```
Comment (cat5inthecradle) [inline aws/ci_builder.rb:38]: I think this is just setting `count` to `1` if it was `0`, which has the effect of ensuring that we always do at least `... -n 1` in the `git log` command below.

Might be cleaner to combine `update_repository` and `log_build_history` into a single function that Fetches, Pulls if necessary, and saves a log of the new commits.

File: aws/ci_builder.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  # Log details of a build to metrics database.
+  # @param [Symbol] environment
+  # @param [String] commit_hash
+  # @param [Symbol] status - :start, :success, :failed
+  def write_build_status(environment, commit_hash, status)
+    event = "#{environment}_#{status}"
+    if environment == :staging
+      Metrics.write_metric(event, commit_hash, Metrics::AUTOMATIC)
+    end
+  end
```
Comment (cat5inthecradle) [inline aws/ci_builder.rb]: This function is defined in both "ci_build" and "ci_builder"

**Review by cat5inthecradle — COMMENTED** [review]
Awesome job making this a pretty straightforward conversion into functions, I would have gotten way too distracted refactoring... Stepping through it, the new code appears to be doing exactly what the old code was. I think the clear function names and steps will make this easier to understand for new engineers. Long-term the tests might be a little brittle, but this code does not change often, so I think we're good.

Have not reviewed test coverage yet.

**Review by cat5inthecradle — APPROVED** [review]
LGTM once we test in an adhoc

File: aws/ci_builder.rb (hunk 1 above)
Comment (pablo-code-org) [reply aws/ci_builder.rb:38]: We had a philosophical discussion about this. 

1) We realized we need to refactor this line. Which is simple enough to be included here but fragile enough to warrant its own pull request.


Once this is checked in, the test cases will be added and I should be able to independently test this too

**Comment by pablo-code-org** [comment]
@cat5inthecradle, all the test and changes have been integrated. I'll be shipping once drone completes

**Comment by pablo-code-org** [comment]
oh and I also added better documentation to the PR

**Pull request merged**